File: 250722/뿌요뿌요/puyo-puyo.js
const fs = require("fs");
const input = fs.readFileSync(0).toString().trim().split('\n');

const n = Number(input[0]);
const grid = input.slice(1, n + 1).map(row => row.split(" ").map(Number));

const visited = Array.from({length: n}, () => Array(n).fill(false));
const dx = [-1, 1, 0 , 0];
const dy = [0, 0, -1, 1];

function dfs(x, y, target){
    let count = 1
    for(let i = 0; i < 4; i++){
        const nx = x + dx[i];
        const ny = y + dy[i];

        if(0<=nx && nx < n && 0<= ny && ny < n && !visited[ny][nx] && target === grid[ny][nx]){
            visited[ny][nx] = true;
            count += dfs(nx, ny, target);
        }
    }

    return count
}
let max = Number.MIN_SAFE_INTEGER
let totalCount = 0
for(let i = 0; i < n; i++){
    for(let j = 0; j < n; j++){
        if(!visited[i][j]){
            visited[i][j] = true;
            const count = dfs(j, i, grid[i][j]);
            if(count >= 4){
                totalCount++
            }
            max = Math.max(count, max);
        }
    }
}

console.log(`${totalCount} ${max}`)

File: 250820/뿌요뿌요/puyo-puyo.py
n = int(input())
grid = [list(map(int, input().split())) for _ in range(n)]

visited = [[False] * n for _ in range(n)]

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

cnt = 0
def dfs(x, y, t):
    global cnt
    for i in range(4):
        nx = dx[i] + x
        ny = dy[i] + y
        if 0 <= nx and nx < n and 0 <= ny and ny < n and grid[ny][nx] == t and not visited[ny][nx]:
            visited[ny][nx] = True
            cnt+=1
            dfs(nx, ny, t)
ans_cnt = 0
max_cnt = 0
for i in range(n):
    for j in range(n):
        if not visited[i][j]:
            cnt = 1
            visited[i][j] = True
            dfs(j, i, grid[i][j])

            if cnt >= 4:
                ans_cnt += 1
            if max_cnt < cnt:
                    max_cnt = cnt

print(ans_cnt, max_cnt)